heudiconv/main.py

```diff
@@ -0,0 +1,340 @@
+    queue : {'SLURM', None}, optional
+        Batch system to submit jobs in parallel. Default is None.
+    queue_args : str or None, optional
+        Additional queue arguments passed as single string of space-separated
+        Argument=Value pairs. Default is None.
+    """
+
+    # To be done asap so anything random is deterministic
+    if random_seed is not None:
+        import random
+        random.seed(random_seed)
+        import numpy
+        numpy.random.seed(random_seed)
+    # Ensure only supported bids options are passed
+    if debug:
+        lgr.setLevel(logging.DEBUG)
+    # Should be possible but only with a single subject -- will be used to
+    # override subject deduced from the DICOMs
+    if files and subjs and len(subjs) > 1:
+        raise ValueError(
+            "Unable to processes multiple `--subjects` with files"
+        )
+
+    if debug:
+        setup_exceptionhook()
+
+    # Deal with provided files or templates
+    # pre-process provided list of files and possibly sort into groups/sessions
+    # Group files per each study/sid/session
+
+    outdir = op.abspath(outdir)
+
+    try:
+        import etelemetry
+        latest = etelemetry.get_project("nipy/heudiconv")
+    except Exception as e:
+        lgr.warning("Could not check for version updates: %s", str(e))
+        latest = {"version": 'Unknown'}
+
+    lgr.info(INIT_MSG(packname=__packagename__,
+                      version=__version__,
+                      latest=latest["version"]))
+
+    if command:
+        process_extra_commands(outdir, command, files, dicom_dir_template,
+                               heuristic, session, subjs, grouping)
+        return
+    #
+    # Load heuristic -- better do it asap to make sure it loads correctly
+    #
+    if not heuristic:
+        raise RuntimeError("No heuristic specified - add to arguments and rerun")
+
+    if queue:
+        lgr.info("Queuing %s conversion", queue)
+        iterarg, iterables = ("files", len(files)) if files else \
+                             ("subjects", len(subjs))
+        queue_conversion(queue, iterarg, iterables, queue_args)
+        sys.exit(0)
+
+    heuristic = load_heuristic(heuristic)
+
+    study_sessions = get_study_sessions(dicom_dir_template, files,
+                                        heuristic, outdir, session,
+                                        subjs, grouping=grouping)
+
+    # extract tarballs, and replace their entries with expanded lists of files
+    # TODO: we might need to sort so sessions are ordered???
+    lgr.info("Need to process %d study sessions", len(study_sessions))
+
+    # processed_studydirs = set()
+
+    for (locator, session, sid), files_or_seqinfo in study_sessions.items():
+
+        # Allow for session to be overloaded from command line
+        if session is not None:
+            session = session
+        if locator is not None:
+            locator = locator
```

Comment:
Smart pycharm greyed out `locator` in the signature of this function which raised my concern.  Here code like
```
        if args.session is not None:
            session = args.session
        if args.locator is not None:
            locator = args.locator
```
was, so `args.` was providing defaults, whenever now loop variables mask away function arguments.  This needs to be fixed up.  It is pity that apparently we have no tests which would have triggered either of those cases!

Reply:
That's a head-scratcher. I combed through the workflow with an eye out for this kind of pattern, but didn't find any other cases.